src/index.ts

```diff
@@ -1,32 +1,81 @@
 import * as utils from "./utils";
+import openAttestationSchema from "./schema/3.0/schema.json";
+import { setData } from "./privacy";
+import * as v2 from "./__generated__/schemaV2";
+import * as v3 from "./__generated__/schemaV3";
+
+const createDocument = (data: any, schemaId?: string) => {
+  const version = openAttestationSchema.$id.split("/")?.[1];
+  if (!version) {
+    throw new Error(`Unable to detect the version of the current open-attestation schema`);
+  }
+  const documentSchema: SchematisedDocument = {
+    // throw error
+    version,
+    data: null
+  };
+  if (schemaId) {
+    documentSchema.schema = schemaId;
+  }
+  return setData(documentSchema, saltData(data));
+};
 
-const createDocument = (data: any, schema: Schema) => {
-  const document = setData({ schema: schema.$id, data: null }, saltData(data));
-  const valid = validate(document, schema);
-  if (valid) {
-    return document;
+class SchemaValidationError extends Error {
+  constructor(message: string, public validationErrors: Ajv.ErrorObject[], public document: any) {
+    super(message);
   }
-  throw new Error(`Invalid document:${JSON.stringify(data, null, 2)}`);
+}
+const isSchemaValidationError = (error: any): error is SchemaValidationError => {
+  return !!error.validationErrors;
 };
 
-export const issueDocument = (data: any, schema: Schema) => {
-  const document = createDocument(data, schema);
-  return sign(document, [digestDocument(document)]);
+interface IssueDocumentOption {
+  externalSchemaId?: string;
+}
+export const issueDocument = (
+  data: unknown,
+  options?: IssueDocumentOption
+): WrappedDocument<v3.OpenAttestationDocument> => {
+  const document: SchematisedDocument = createDocument(data, options?.externalSchemaId);
+  const errors = validate(document, getSchema("open-attestation/3.0"));
+  if (errors.length > 0) {
+    throw new SchemaValidationError("Invalid document", errors, document);
+  }
+  return wrap(document, [digestDocument(document)]);
 };
 
-export const issueDocuments = (dataArray: any[], schema: Schema) => {
-  const documents = dataArray.map(data => createDocument(data, schema));
+export const issueDocuments = (
+  dataArray: unknown[],
+  options?: IssueDocumentOption
+): WrappedDocument<v3.OpenAttestationDocument>[] => {
+  const documents = dataArray.map(data => createDocument(data, options?.externalSchemaId));
+  documents.forEach(document => {
+    const errors = validate(document, getSchema("open-attestation/3.0"));
+    if (errors.length > 0) {
+      throw new SchemaValidationError("Invalid document", errors, document);
+    }
+  });
+
   const batchHashes = documents.map(digestDocument);
-  return documents.map(doc => sign(doc, batchHashes));
+  return documents.map(doc => wrap(doc, batchHashes));
+};
+
+export const validateSchema = (document: WrappedDocument): boolean => {
+  return validate(document, getSchema(`open-attestation/${document?.version || "2.0"}`)).length === 0;
```

Comment:
actually I'm thinking instead of storing `3.0` in the version, does it make sense to store the version as `open-attestation/3.0` so that it is unambiguous which version (OA schema or their subschema)? Dont hit me.